File: weechat-android/src/main/java/android/support/v7/preference/FontPreference.java
package android.support.v7.preference;

import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckedTextView;

import com.ubergeek42.WeechatAndroid.utils.Constants;

import java.util.Collections;
import java.util.LinkedList;

public class FontPreference extends DialogPreference {

    ////////////////////////////////////////////////////////////////////////////////////////////////

    public FontPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public @NonNull String getFontPath() {
        return getSharedPreferences().getString(getKey(), Constants.PREF_BUFFER_FONT_D);
    }

    public void setFontPath(@NonNull String path) {
        getSharedPreferences().edit().putString(getKey(), path).commit();
        notifyChanged();
    }

    @Override public CharSequence getSummary() {
        StringBuilder sb = new StringBuilder();
        sb.append("Non-monospace fonts will not work well with alignment.");
        sb.append("\n\nSearch Path:");
        for (String p: FontManager.FONT_DIRS)
            sb.append("\n    ").append(p);
        sb.append("\n\nCurrent Value:\n    ");
        sb.append("".equals(getFontPath()) ? "Default" : getFontPath());
        return sb.toString();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    public static class FontPreferenceFragment extends PreferenceDialogFragmentCompat implements DialogInterface.OnClickListener {

        private LinkedList<FontManager.FontInfo> fonts;
        private LayoutInflater inflater;

        public static FontPreferenceFragment newInstance(String key) {
            FontPreferenceFragment fragment = new FontPreferenceFragment();
            Bundle b = new Bundle(1);
            b.putString("key", key);
            fragment.setArguments(b);
            return fragment;
        }

        protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
            super.onPrepareDialogBuilder(builder);

            inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            fonts = FontManager.enumerateFonts();
            Collections.sort(fonts);

            // add a "fake" default monospace font
            fonts.addFirst(new FontManager.FontInfo("Default", "", Typeface.MONOSPACE));

            // get index of currently selected font
            String currentPath = ((FontPreference) getPreference()).getFontPath();
            int idx = 0, checked_item = 0;
            for (FontManager.FontInfo font : fonts) {
                if (font.path.equals(currentPath)) {checked_item = idx; break;}
                idx++;
            }

            builder.setSingleChoiceItems(new FontAdapter(), checked_item, this);
            builder.setPositiveButton(null, null);
        }

        public void onClick(DialogInterface dialog, int which) {
            if (which >= 0)
                ((FontPreference) getPreference()).setFontPath(fonts.get(which).path);
            dialog.dismiss();
        }

        @Override public void onDialogClosed(boolean b) {}

        ////////////////////////////////////////////////////////////////////////////////////////////

        public class FontAdapter extends BaseAdapter {
            @Override public int getCount() {
                return fonts.size();
            }

            @Override public Object getItem(int position) {
                return fonts.get(position);
            }

            @Override public long getItemId(int position) {
                return position;
            }

            @Override public View getView(int position, View view, ViewGroup parent) {
                if (view == null) {
                    view = inflater.inflate(android.support.v7.appcompat.R.layout.select_dialog_singlechoice_material, parent, false);
                    CheckedTextView tv = (CheckedTextView) view.findViewById(android.R.id.text1);
                    tv.setEllipsize(TextUtils.TruncateAt.END);
                    tv.setSingleLine();
                }

                FontManager.FontInfo font = (FontManager.FontInfo) getItem(position);
                CheckedTextView tv = (CheckedTextView) view.findViewById(android.R.id.text1);
                tv.setTypeface(font.typeface);
                tv.setText(font.name);
                return view;
            }
        }
    }
}

File: app/src/main/java/androidx/preference/FontPreference.kt
package androidx.preference

import android.content.Context
import android.content.DialogInterface
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.CheckedTextView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import com.ubergeek42.WeechatAndroid.R
import com.ubergeek42.WeechatAndroid.utils.Constants
import java.io.File
import androidx.core.content.edit


class FontPreference(context: Context, attrs: AttributeSet?) : DialogPreference(context, attrs), DialogFragmentGetter {
    private var fontPaths: Set<String>?
        get() = sharedPreferences!!.getStringSet(Constants.PREF_BUFFER_FONTS, Constants.PREF_BUFFER_FONTS_D)
        set(paths) {
            sharedPreferences!!.edit { putStringSet(Constants.PREF_BUFFER_FONTS, paths) }
            notifyChanged()
        }

    override fun getSummary(): CharSequence {
        val paths = fontPaths
        return if (paths.isNullOrEmpty()) {
            context.getString(R.string.pref__FontPreference__default)
        } else {
            paths.joinToString(", ") { File(it).name }
        }
    }

    override fun getDialogFragment(): DialogFragment {
        return FontPreferenceFragment()
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    class FontPreferenceFragment : PreferenceDialogFragmentCompat(), DialogInterface.OnClickListener {
        private lateinit var typefaces: List<TypefaceInfo>
        private lateinit var inflater: LayoutInflater

        @OptIn(ExperimentalStdlibApi::class)
        override fun onPrepareDialogBuilder(builder: AlertDialog.Builder) {
            super.onPrepareDialogBuilder(builder)
            inflater = LayoutInflater.from(context)

            val defaultTypeface = TypefaceInfo.getDefault(requireContext().getString(R.string.pref__FontPreference__default))
            val availableTypefaces = enumerateTypefaces(requireContext())
            typefaces = listOf(defaultTypeface) + availableTypefaces.sortedBy { it.name.lowercase() }

            typefaces.forEach {
                println("Typeface: ${it.name} ${it.fonts}")
            }

            val currentPaths = (preference as FontPreference).fontPaths
            val currentIndex = typefaces.indexOfFirst { currentPaths == it.fontFilePaths.toSet() } // -1 is ok

            builder.setSingleChoiceItems(FontAdapter(), currentIndex, this)
            builder.setPositiveButton(getString(R.string.pref__FontPreference__import_button)) { _, _ ->
                FontManager.requestFontImport(requireActivity())
                dismiss()
            }
        }

        override fun onClick(dialog: DialogInterface, which: Int) {
            if (which >= 0) (preference as FontPreference).fontPaths = typefaces[which].fontFilePaths.toSet()
            dialog.dismiss()
        }

        override fun onDialogClosed(b: Boolean) {}

        ////////////////////////////////////////////////////////////////////////////////////////////

        private inner class FontAdapter : BaseAdapter() {
            override fun getCount() = typefaces.size
            override fun getItem(position: Int) = typefaces[position]
            override fun getItemId(position: Int) = position.toLong()

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val view = convertView ?: inflater.inflate(androidx.appcompat.R.layout.select_dialog_singlechoice_material, parent, false)
                val textView = view.findViewById<CheckedTextView>(android.R.id.text1)

                val fontInfo = getItem(position)
                textView.apply {
                    ellipsize = TextUtils.TruncateAt.END
                    //setSingleLine()
                    typeface = fontInfo.typeface
                    text = fontInfo.name + " (${fontInfo.getStylesDescription().joinToString(", ")})"
                }

                return view
            }
        }
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        val summary = holder.findViewById(android.R.id.summary) as TextView
        summary.maxHeight = Int.MAX_VALUE
    }
}